internal/controller/apiextensions/composite/composed.go

```diff
@@ -54,6 +60,224 @@ const (
+func (fn CompositionTemplateAssociatorFn) AssociateTemplates(ctx context.Context, cr resource.Composite, comp *v1.Composition) ([]TemplateAssociation, error) {
+	return fn(ctx, cr, comp)
+}
+
+// A GarbageCollectingAssociator associates a Composition's resource templates
+// with (references to) composed resources. It tries to associate them by
+// checking the template name annotation of each referenced resource. If any
+// template or existing composed resource can't be associated by name it falls
+// back to associating them by order. If it encounters a referenced resource
+// that corresponds to a non-extistent template the resource will be garbage
+// collected (i.e. deleted).
+type GarbageCollectingAssociator struct {
+	client client.Client
+}
+
+// NewGarbageCollectingAssociator returns a CompositionTemplateAssociator that
+// may garbage collect composed resources.
+func NewGarbageCollectingAssociator(c client.Client) *GarbageCollectingAssociator {
+	return &GarbageCollectingAssociator{client: c}
+}
+
+// AssociateTemplates with composed resources.
+func (a *GarbageCollectingAssociator) AssociateTemplates(ctx context.Context, cr resource.Composite, comp *v1.Composition) ([]TemplateAssociation, error) { //nolint:gocyclo
+	// NOTE(negz): This method is a little over our complexity goal. Be wary of
+	// making it more complex.
+
+	templates := map[string]int{}
+	for i, t := range comp.Spec.Resources {
+		if t.TemplateName == nil {
+			// If our templates aren't named we fall back to assuming that the
+			// existing resource reference array (if any) already matches the
+			// order of our resource template array.
+			return AssociateByOrder(comp.Spec.Resources, cr.GetResourceReferences()), nil
+		}
+		templates[*t.TemplateName] = i
+	}
+
+	tas := make([]TemplateAssociation, len(comp.Spec.Resources))
+	for i := range comp.Spec.Resources {
+		tas[i] = TemplateAssociation{Template: comp.Spec.Resources[i]}
+	}
+
+	for _, ref := range cr.GetResourceReferences() {
+		cd := composed.New(composed.FromReference(ref))
+		nn := types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}
+		err := a.client.Get(ctx, nn, cd)
+
+		// We believe we created this resource, but it no longer exists.
+		if kerrors.IsNotFound(err) {
+			continue
+		}
+
+		if err != nil {
+			return nil, errors.Wrap(err, errGetComposed)
+		}
+
+		name := GetCompositionTemplateName(cd)
+		if name == "" {
+			// All of our templates are named, but this existing composed
+			// resource is not associated with a named template. It's likely
+			// that our Composition was just migrated from anonymous to named
+			// templates. We fall back to assuming that the existing resource
+			// reference array already matches the order of our resource
+			// template array. Existing composed resources should be annotated
+			// at render time with the name of the template used to create them.
+			return AssociateByOrder(comp.Spec.Resources, cr.GetResourceReferences()), nil
```

Comment:
This could also happen if the Composition author decided to patch to `metadata.annotations`. This replaces rather than merges annotations, and would therefore wipe this one out. We've seen folks accidentally wipe out their external name annotation by doing this. I think we should document this behaviour and warn folks to be careful when patching object metadata.

I initially considered keeping the association between templates and composed resources within the XR's `resourceRefs` array. This approach wouldn't be prone to this issue. It would also avoid the need to `Get` the composed resources (which is less of an issue since we're just reading from cache). Unfortunately I couldn't see a clean way to make the API change. i.e.:

```yaml
# This violates API conventions - refs must be a subset of v1.ObjectReference
resourceRefs:
- templateName: rdsinstance
  apiVersion: database.aws.crossplane.io/v1beta1
  kind: RDSInstance
  name: example-amd92
```

```yaml
# This is a breaking API change
resourceRefs:
- templateName: rdsinstance
  ref:
    apiVersion: database.aws.crossplane.io/v1beta1
    kind: RDSInstance
    name: example-amd92
```

Reply:
Hrm, maybe. Unless we want to try to support merge patching simple maps?

Reply:
Yeah that's an option, too.